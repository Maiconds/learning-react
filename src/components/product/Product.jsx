import React from 'react';
import PropTypes from 'prop-types';

const Product = ({ name, price, description, showPrice }) => (
  <div>
    <p className="product-name">{name}</p>
    { showPrice ? <p className="product-price">{price}</p> : null}
    <p className="product-description">{description}</p>
  </div>
);

Product.propTypes = {
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  description: PropTypes.string,
  showPrice: PropTypes.bool.isRequired
};

export default Product; 
