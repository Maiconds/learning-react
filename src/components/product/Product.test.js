import React from 'react';
import { shallow } from 'enzyme';
import Product from './Product';

describe('Testing Product rendering', () => {
  
  describe('All props', () => {
    const productName = 'Rice';
    const productPrice = 2.25;
    const productDescription = '500g';

    const wrapper = shallow(<Product name={productName} price={productPrice} description={productDescription} showPrice={true} />);

    it('Should render product name', () => {
      expect(wrapper.find('.product-name')).toHaveLength(1);
      expect(wrapper.find('.product-name').props().children).toBe(productName);
    });

    it('Should render product price', () => {
      expect(wrapper.find('.product-price')).toHaveLength(1);
      expect(wrapper.find('.product-price').props().children).toBe(productPrice);
    });

    it('Should render product description', () => {
      expect(wrapper.find('.product-description')).toHaveLength(1);
      expect(wrapper.find('.product-description').props().children).toBe(productDescription);
    });

    it('Should mactch the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

  });

  describe('No description', () => {
    const productName = 'Rice';
    const productPrice = 2.25;

    const wrapper = shallow(<Product name={productName} price={productPrice} showPrice={true} />);

    it('Should render product name', () => {
      expect(wrapper.find('.product-name')).toHaveLength(1);
      expect(wrapper.find('.product-name').props().children).toBe(productName);
    });

    it('Should render product price', () => {
      expect(wrapper.find('.product-price')).toHaveLength(1);
      expect(wrapper.find('.product-price').props().children).toBe(productPrice);
    });

    it('Should not render product description', () => {
      expect(wrapper.find('.product-description')).toHaveLength(1);
      expect(wrapper.find('.product-description').props().children).toBeUndefined();
    });

    it('Should mactch the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

  });

  describe('Hiding price', () => {
    const productName = 'Rice';
    const productPrice = 2.25;
    const productDescription = '500g';

    const wrapper = shallow(<Product name={productName} price={productPrice} description={productDescription} showPrice={false} />);

    it('Should render product name', () => {
      expect(wrapper.find('.product-name')).toHaveLength(1);
      expect(wrapper.find('.product-name').props().children).toBe(productName);
    });

    it('Should not render product price', () => {
      expect(wrapper.find('.product-price')).toHaveLength(0);
    });

    it('Should render product description', () => {
      expect(wrapper.find('.product-description')).toHaveLength(1);
      expect(wrapper.find('.product-description').props().children).toBe(productDescription);
    });

    it('Should mactch the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

  });

});