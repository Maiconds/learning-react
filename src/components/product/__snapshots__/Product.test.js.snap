// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Product rendering All props Should mactch the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    description="500g"
    name="Rice"
    price={2.25}
    showPrice={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="product-name"
        >
          Rice
        </p>,
        <p
          className="product-price"
        >
          2.25
        </p>,
        <p
          className="product-description"
        >
          500g
        </p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Rice",
          "className": "product-name",
        },
        "ref": null,
        "rendered": "Rice",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 2.25,
          "className": "product-price",
        },
        "ref": null,
        "rendered": 2.25,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "500g",
          "className": "product-description",
        },
        "ref": null,
        "rendered": "500g",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="product-name"
          >
            Rice
          </p>,
          <p
            className="product-price"
          >
            2.25
          </p>,
          <p
            className="product-description"
          >
            500g
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Rice",
            "className": "product-name",
          },
          "ref": null,
          "rendered": "Rice",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 2.25,
            "className": "product-price",
          },
          "ref": null,
          "rendered": 2.25,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "500g",
            "className": "product-description",
          },
          "ref": null,
          "rendered": "500g",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing Product rendering Hiding price Should mactch the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    description="500g"
    name="Rice"
    price={2.25}
    showPrice={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="product-name"
        >
          Rice
        </p>,
        null,
        <p
          className="product-description"
        >
          500g
        </p>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Rice",
          "className": "product-name",
        },
        "ref": null,
        "rendered": "Rice",
        "type": "p",
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "500g",
          "className": "product-description",
        },
        "ref": null,
        "rendered": "500g",
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="product-name"
          >
            Rice
          </p>,
          null,
          <p
            className="product-description"
          >
            500g
          </p>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Rice",
            "className": "product-name",
          },
          "ref": null,
          "rendered": "Rice",
          "type": "p",
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "500g",
            "className": "product-description",
          },
          "ref": null,
          "rendered": "500g",
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Testing Product rendering No description Should mactch the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Product
    name="Rice"
    price={2.25}
    showPrice={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="product-name"
        >
          Rice
        </p>,
        <p
          className="product-price"
        >
          2.25
        </p>,
        <p
          className="product-description"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Rice",
          "className": "product-name",
        },
        "ref": null,
        "rendered": "Rice",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 2.25,
          "className": "product-price",
        },
        "ref": null,
        "rendered": 2.25,
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "product-description",
        },
        "ref": null,
        "rendered": null,
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="product-name"
          >
            Rice
          </p>,
          <p
            className="product-price"
          >
            2.25
          </p>,
          <p
            className="product-description"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Rice",
            "className": "product-name",
          },
          "ref": null,
          "rendered": "Rice",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 2.25,
            "className": "product-price",
          },
          "ref": null,
          "rendered": 2.25,
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "product-description",
          },
          "ref": null,
          "rendered": null,
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
